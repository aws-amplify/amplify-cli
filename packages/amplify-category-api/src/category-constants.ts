import { AmplifyCategories } from 'amplify-cli-core';

export const ADMIN_QUERIES_NAME = 'AdminQueries';
export const NETWORK_STACK_LOGICAL_ID = 'NetworkStack';
export const category = AmplifyCategories.API;
