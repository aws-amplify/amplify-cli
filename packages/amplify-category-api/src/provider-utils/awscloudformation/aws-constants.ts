import * as path from 'path';

export const provider = 'awscloudformation';
export const parametersFileName = 'api-params.json';
export const cfnParametersFilename = 'parameters.json';
export const gqlSchemaFilename = 'schema.graphql';

export const rootAssetDir = path.resolve(path.join(__dirname, '..', '..', '..', 'resources', 'awscloudformation'));
