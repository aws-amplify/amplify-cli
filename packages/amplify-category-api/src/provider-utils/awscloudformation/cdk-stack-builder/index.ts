export * from './apigw-stack-builder';
export * from './apigw-stack-transform';
export * from './types';
