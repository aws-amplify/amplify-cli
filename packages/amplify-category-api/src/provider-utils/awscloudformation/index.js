const path = require('path');
const fs = require('fs-extra');

const serviceMetadataFor = service => require('../supported-services').supportedServices[service];
const datasourceMetadataFor = datasource => require('../supported-datasources').supportedDatasources[datasource];
const getServiceWalkthrough = walkthroughFilename => require(`./service-walkthroughs/${walkthroughFilename}`)['serviceWalkthrough'];

const parametersFileName = 'api-params.json';
const cfnParametersFilename = 'parameters.json';

const rootAssetDir = path.resolve(path.join(__dirname, '../../../resources/awscloudformation'));

function copyCfnTemplate(context, category, options, cfnFilename) {
  const { amplify } = context;
  const targetDir = amplify.pathManager.getBackendDirPath();

  const copyJobs = [
    {
      dir: path.join(rootAssetDir, 'cloudformation-templates'),
      template: cfnFilename,
      target: `${targetDir}/${category}/${options.resourceName}/${options.resourceName}-cloudformation-template.json`,
    },
  ];

  // copy over the files
  return context.amplify.copyBatch(context, copyJobs, options, true, false);
}

function console(context, service) {
  const { serviceWalkthroughFilename } = serviceMetadataFor(service);
  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
  const { openConsole } = require(serviceWalkthroughSrc);

  if (!openConsole) {
    context.print.error('Opening console functionality not available for this option');
    process.exit(0);
  }

  return openConsole(context);
}

function addResource(context, category, service, options) {
  let answers;
  const serviceMetadata = serviceMetadataFor(service);
  let { cfnFilename } = serviceMetadata;
  const { defaultValuesFilename, serviceWalkthroughFilename } = serviceMetadata;
  const projectBackendDirPath = context.amplify.pathManager.getBackendDirPath();

  return getServiceWalkthrough(serviceWalkthroughFilename)(context, defaultValuesFilename, serviceMetadata).then(result => {
    if (result.answers) {
      ({ answers } = result);
      options.dependsOn = result.dependsOn;
    } else {
      answers = result;
    }
    if (result.output) {
      options.output = result.output;
    }
    if (!result.noCfnFile) {
      if (answers.customCfnFile) {
        cfnFilename = answers.customCfnFile;
      }
      copyCfnTemplate(context, category, answers, cfnFilename);

      const parameters = { ...answers };
      const cfnParameters = {
        authRoleName: {
          Ref: 'AuthRoleName',
        },
        unauthRoleName: {
          Ref: 'UnauthRoleName',
        },
      };
      const resourceDirPath = path.join(projectBackendDirPath, category, parameters.resourceName);
      fs.ensureDirSync(resourceDirPath);

      const parametersFilePath = path.join(resourceDirPath, parametersFileName);
      let jsonString = JSON.stringify(parameters, null, 4);
      fs.writeFileSync(parametersFilePath, jsonString, 'utf8');

      const cfnParametersFilePath = path.join(resourceDirPath, cfnParametersFilename);
      jsonString = JSON.stringify(cfnParameters, null, 4);
      fs.writeFileSync(cfnParametersFilePath, jsonString, 'utf8');
    }
    context.amplify.updateamplifyMetaAfterResourceAdd(category, answers.resourceName, options);
    return answers.resourceName;
  });
}

async function updateResource(context, category, service) {
  let answers;
  const serviceMetadata = serviceMetadataFor(service);
  let { cfnFilename } = serviceMetadata;
  const { defaultValuesFilename, serviceWalkthroughFilename } = serviceMetadata;
  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
  const { updateWalkthrough } = require(serviceWalkthroughSrc);
  const projectBackendDirPath = context.amplify.pathManager.getBackendDirPath();

  if (!updateWalkthrough) {
    context.print.error('Update functionality not available for this option');
    process.exit(0);
  }

  return updateWalkthrough(context, defaultValuesFilename, serviceMetadata).then(result => {
    const options = {};
    if (result) {
      if (result.answers) {
        ({ answers } = result);
        options.dependsOn = result.dependsOn;
      } else {
        answers = result;
      }

      if (!result.noCfnFile) {
        if (answers.customCfnFile) {
          cfnFilename = answers.customCfnFile;
        }
        copyCfnTemplate(context, category, answers, cfnFilename);
        const parameters = { ...answers };
        const resourceDirPath = path.join(projectBackendDirPath, category, parameters.resourceName);
        fs.ensureDirSync(resourceDirPath);
        const parametersFilePath = path.join(resourceDirPath, parametersFileName);
        const jsonString = JSON.stringify(parameters, null, 4);
        fs.writeFileSync(parametersFilePath, jsonString, 'utf8');
        context.amplify.updateamplifyMetaAfterResourceUpdate(category, answers.resourceName, 'dependsOn', answers.dependsOn);
      }
    }
  });
}

async function migrateResource(context, projectPath, service, resourceName) {
  const serviceMetadata = serviceMetadataFor(service);
  const { serviceWalkthroughFilename } = serviceMetadata;
  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
  const { migrate } = require(serviceWalkthroughSrc);

  if (!migrate) {
    context.print.info(`No migration required for ${resourceName}`);
    return;
  }

  return await migrate(context, projectPath, resourceName);
}

function addDatasource(context, category, datasource) {
  const serviceMetadata = datasourceMetadataFor(datasource);
  const { defaultValuesFilename, serviceWalkthroughFilename } = serviceMetadata;
  return getServiceWalkthrough(serviceWalkthroughFilename)(context, defaultValuesFilename, serviceMetadata);
}

function getPermissionPolicies(context, service, resourceName, crudOptions) {
  const serviceMetadata = serviceMetadataFor(service);
  const { serviceWalkthroughFilename } = serviceMetadata;
  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
  const { getIAMPolicies } = require(serviceWalkthroughSrc);

  if (!getIAMPolicies) {
    context.print.info(`No policies found for ${resourceName}`);
    return;
  }

  return getIAMPolicies(resourceName, crudOptions);
}

module.exports = {
  addResource,
  updateResource,
  console,
  migrateResource,
  addDatasource,
  getPermissionPolicies,
  rootAssetDir,
};
