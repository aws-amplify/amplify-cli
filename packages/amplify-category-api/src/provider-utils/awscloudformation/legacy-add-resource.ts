import { $TSAny, $TSContext, $TSObject, isResourceNameUnique, JSONUtilities, pathManager } from 'amplify-cli-core';
import * as fs from 'fs-extra';
import * as path from 'path';
import { cfnParametersFilename, parametersFileName, rootAssetDir } from './aws-constants';
import { serviceMetadataFor } from './utils/dynamic-imports';

// this is the old logic for generating resources in the project directory
// it is still used for adding REST APIs
export const legacyAddResource = async (
  serviceWalkthroughPromise: Promise<$TSAny>,
  context: $TSContext,
  category: string,
  service: string,
  options: $TSObject,
) => {
  let answers;
  let { cfnFilename } = await serviceMetadataFor(service);

  const result = await serviceWalkthroughPromise;

  if (result.answers) {
    ({ answers } = result);
    options.dependsOn = result.dependsOn;
  } else {
    answers = result;
  }
  if (result.output) {
    options.output = result.output;
  }
  if (!result.noCfnFile) {
    if (answers.customCfnFile) {
      cfnFilename = answers.customCfnFile;
    }
    addPolicyResourceNameToPaths(answers.paths);
    copyCfnTemplate(context, category, answers, cfnFilename);

    const parameters = { ...answers };
    const resourceDirPath = pathManager.getResourceDirectoryPath(undefined, category, parameters.resourceName);

    isResourceNameUnique(category, parameters.resourceName);

    fs.ensureDirSync(resourceDirPath);

    const parametersFilePath = path.join(resourceDirPath, parametersFileName);
    JSONUtilities.writeJson(parametersFilePath, parameters);

    const cfnParametersFilePath = path.join(resourceDirPath, cfnParametersFilename);
    JSONUtilities.writeJson(cfnParametersFilePath, {});
  }
  context.amplify.updateamplifyMetaAfterResourceAdd(category, answers.resourceName, options);
  return answers.resourceName;
};

// exported because the update flow still uses this method directly for now
export const copyCfnTemplate = (context: $TSContext, category: string, options, cfnFilename) => {
  const resourceDirPath = pathManager.getResourceDirectoryPath(undefined, category, options.resourceName);

  const copyJobs = [
    {
      dir: path.join(rootAssetDir, 'cloudformation-templates'),
      template: cfnFilename,
      target: path.join(resourceDirPath, `${options.resourceName}-cloudformation-template.json`),
    },
  ];

  // copy over the files
  return context.amplify.copyBatch(context, copyJobs, options, true, false);
};

export const addPolicyResourceNameToPaths = paths => {
  if (Array.isArray(paths)) {
    paths.forEach(p => {
      const pathName = p.name;
      if (typeof pathName === 'string') {
        p.policyResourceName = pathName.replace(/{[a-zA-Z0-9\-]+}/g, '*');
      }
    });
  }
};
