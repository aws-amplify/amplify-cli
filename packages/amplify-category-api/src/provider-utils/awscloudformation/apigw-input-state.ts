import {
  $TSContext,
  $TSObject,
  AmplifyCategories,
  AmplifySupportedService,
  CLIInputSchemaValidator,
  getMigrateResourceMessageForOverride,
  isResourceNameUnique,
  JSONUtilities,
  PathConstants,
  pathManager,
  stateManager,
} from 'amplify-cli-core';
import { printer, prompter } from 'amplify-prompts';
import * as fs from 'fs-extra';
import { join } from 'path';
import { ApigwInputs, ApigwStackTransform, CrudOperation, Path, PermissionSetting } from './cdk-stack-builder';
import { ApigwWalkthroughReturnPromise } from './service-walkthrough-types/apigw-types';

export class ApigwInputState {
  projectRootPath: string;
  resourceName: string;
  paths: { [pathName: string]: Path };

  constructor(private readonly context: $TSContext, resourceName?: string) {
    this.projectRootPath = pathManager.findProjectRoot();
    this.resourceName = resourceName;
  }

  public addAdminQueriesResource = async (adminQueriesProps: AdminQueriesProps) => {
    this.resourceName = adminQueriesProps.apiName;
    this.paths = {
      '/{proxy+}': {
        lambdaFunction: adminQueriesProps.functionName,
        permissions: {
          setting: PermissionSetting.PRIVATE,
          auth: [CrudOperation.CREATE, CrudOperation.READ, CrudOperation.UPDATE, CrudOperation.DELETE],
        },
      },
    };

    await this.createApigwArtifacts();

    // Update amplify-meta and backend-config
    const backendConfigs = {
      service: AmplifySupportedService.APIGW,
      providerPlugin: 'awscloudformation',
      authorizationType: 'AMAZON_COGNITO_USER_POOLS',
      dependsOn: adminQueriesProps.dependsOn,
    };

    await this.context.amplify.updateamplifyMetaAfterResourceAdd(AmplifyCategories.API, adminQueriesProps.apiName, backendConfigs);
  };

  public updateAdminQueriesResource = async (adminQueriesProps: AdminQueriesProps) => {
    this.resourceName = adminQueriesProps.apiName;
    this.paths = {
      '/{proxy+}': {
        lambdaFunction: adminQueriesProps.functionName,
        permissions: {
          setting: PermissionSetting.PRIVATE,
          auth: [CrudOperation.CREATE, CrudOperation.READ, CrudOperation.UPDATE, CrudOperation.DELETE],
        },
      },
    };

    await this.createApigwArtifacts();

    await this.context.amplify.updateamplifyMetaAfterResourceUpdate(
      AmplifyCategories.API,
      adminQueriesProps.apiName,
      'dependsOn',
      adminQueriesProps.dependsOn,
    );
  };

  public addApigwResource = async (serviceWalkthroughPromise: ApigwWalkthroughReturnPromise, options: $TSObject) => {
    const { answers } = await serviceWalkthroughPromise;

    this.resourceName = answers.resourceName;
    this.paths = answers.paths;
    options.dependsOn = answers.dependsOn;

    isResourceNameUnique(AmplifyCategories.API, this.resourceName);

    await this.createApigwArtifacts();

    this.context.amplify.updateamplifyMetaAfterResourceAdd(AmplifyCategories.API, this.resourceName, options);
    return this.resourceName;
  };

  public updateApigwResource = async (updateWalkthroughPromise: Promise<$TSObject>) => {
    const { answers } = await updateWalkthroughPromise;

    this.resourceName = answers.resourceName;
    this.paths = answers.paths;

    await this.createApigwArtifacts();

    this.context.amplify.updateamplifyMetaAfterResourceUpdate(AmplifyCategories.API, this.resourceName, 'dependsOn', answers.dependsOn);
    return this.resourceName;
  };

  public migrateAdminQueries = async (adminQueriesProps: AdminQueriesProps) => {
    this.resourceName = this.resourceName ?? adminQueriesProps.apiName;
    if (!(await prompter.yesOrNo(getMigrateResourceMessageForOverride(AmplifyCategories.API, this.resourceName, true), true))) {
      return;
    }
    const resourceDirPath = pathManager.getResourceDirectoryPath(this.projectRootPath, AmplifyCategories.API, this.resourceName);

    fs.removeSync(join(resourceDirPath, PathConstants.ParametersJsonFileName));
    fs.removeSync(join(resourceDirPath, 'admin-queries-cloudformation-template.json'));

    return this.updateAdminQueriesResource(adminQueriesProps);
  };

  public migrateApigwResource = async (resourceName: string) => {
    this.resourceName = this.resourceName ?? resourceName;
    if (!(await prompter.yesOrNo(getMigrateResourceMessageForOverride(AmplifyCategories.API, this.resourceName, true), true))) {
      return;
    }
    const deprecatedParametersFileName = 'api-params.json';
    const resourceDirPath = pathManager.getResourceDirectoryPath(this.projectRootPath, AmplifyCategories.API, this.resourceName);
    const deprecatedParametersFilePath = join(resourceDirPath, deprecatedParametersFileName);
    let deprecatedParameters: $TSObject;
    try {
      deprecatedParameters = JSONUtilities.readJson<$TSObject>(deprecatedParametersFilePath);
    } catch (e) {
      printer.error(`Error reading ${deprecatedParametersFileName} file for ${this.resourceName} resource`);
      throw e;
    }

    this.paths = {};

    function _convertDeprecatedPermissionStringToCRUD(deprecatedPrivacy: string) {
      let privacyList: string[];
      if (deprecatedPrivacy === 'r') {
        privacyList = [CrudOperation.READ];
      } else if (deprecatedPrivacy === 'rw') {
        privacyList = [CrudOperation.CREATE, CrudOperation.READ, CrudOperation.UPDATE, CrudOperation.DELETE];
      }
      return privacyList;
    }

    function _convertDeprecatedPermissionArrayToCRUD(deprecatedPrivacyArray: string[]): CrudOperation[] {
      const opMap: Record<string, CrudOperation> = {
        '/POST': CrudOperation.CREATE,
        '/GET': CrudOperation.READ,
        '/PUT': CrudOperation.UPDATE,
        '/PATCH': CrudOperation.UPDATE,
        '/DELETE': CrudOperation.DELETE,
      };
      return Array.from(new Set(deprecatedPrivacyArray.map(op => opMap[op])));
    }

    if (!Array.isArray(deprecatedParameters.paths) || deprecatedParameters.paths.length < 1) {
      throw new Error(`Expected paths to be defined in "${deprecatedParametersFilePath}", but none found.`);
    }

    deprecatedParameters.paths.forEach((path: $TSObject) => {
      let pathPermissionSetting =
        path.privacy?.open === true
          ? PermissionSetting.OPEN
          : path.privacy?.private === true
          ? PermissionSetting.PRIVATE
          : PermissionSetting.PROTECTED;

      let auth;
      let guest;
      let groups;
      // convert deprecated permissions to CRUD structure
      if (typeof path.privacy?.auth === 'string' && ['r', 'rw'].includes(path.privacy.auth)) {
        auth = _convertDeprecatedPermissionStringToCRUD(path.privacy.auth);
      } else if (Array.isArray(path.privacy?.auth)) {
        auth = _convertDeprecatedPermissionArrayToCRUD(path.privacy.auth);
      }

      if (typeof path.privacy?.unauth === 'string' && ['r', 'rw'].includes(path.privacy.unauth)) {
        guest = _convertDeprecatedPermissionStringToCRUD(path.privacy.unauth);
      } else if (Array.isArray(path.privacy?.unauth)) {
        guest = _convertDeprecatedPermissionArrayToCRUD(path.privacy.unauth);
      }

      if (path.privacy?.userPoolGroups) {
        groups = {};
        for (const [userPoolGroupName, crudOperations] of Object.entries(path.privacy.userPoolGroups)) {
          if (typeof crudOperations === 'string' && ['r', 'rw'].includes(crudOperations)) {
            groups[userPoolGroupName] = _convertDeprecatedPermissionStringToCRUD(crudOperations);
          } else if (Array.isArray(crudOperations)) {
            groups[userPoolGroupName] = _convertDeprecatedPermissionArrayToCRUD(crudOperations);
          }
        }
      }

      this.paths[path.name] = {
        permissions: {
          setting: pathPermissionSetting,
          auth,
          guest,
          groups,
        },
        lambdaFunction: path.lambdaFunction,
      };
    });

    fs.removeSync(deprecatedParametersFilePath);
    fs.removeSync(join(resourceDirPath, PathConstants.ParametersJsonFileName));
    fs.removeSync(join(resourceDirPath, `${this.resourceName}-cloudformation-template.json`));

    await this.createApigwArtifacts();
  };

  public cliInputsFileExists() {
    return stateManager.resourceInputsJsonExists(this.projectRootPath, AmplifyCategories.API, this.resourceName);
  }

  public getCliInputPayload() {
    return stateManager.getResourceInputsJson(this.projectRootPath, AmplifyCategories.API, this.resourceName);
  }

  public isCLIInputsValid(cliInputs?: ApigwInputs) {
    if (!cliInputs) {
      cliInputs = this.getCliInputPayload();
    }

    const schemaValidator = new CLIInputSchemaValidator(AmplifySupportedService.APIGW, AmplifyCategories.API, 'APIGatewayCLIInputs');
    schemaValidator.validateInput(JSONUtilities.stringify(cliInputs));
  }

  private async createApigwArtifacts() {
    const resourceDirPath = pathManager.getResourceDirectoryPath(this.projectRootPath, AmplifyCategories.API, this.resourceName);
    fs.ensureDirSync(resourceDirPath);

    const buildDirPath = join(resourceDirPath, PathConstants.BuildDirName);
    fs.ensureDirSync(buildDirPath);

    stateManager.setResourceInputsJson(this.projectRootPath, AmplifyCategories.API, this.resourceName, { version: 1, paths: this.paths });

    stateManager.setResourceParametersJson(this.projectRootPath, AmplifyCategories.API, this.resourceName, {});

    const stack = new ApigwStackTransform(this.context, this.resourceName, this);
    await stack.transform();
  }
}

export type AdminQueriesProps = {
  apiName: string;
  functionName: string;
  authResourceName: string;
  dependsOn: $TSObject[];
};
