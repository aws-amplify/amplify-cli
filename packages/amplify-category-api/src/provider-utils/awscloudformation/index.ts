import { $TSAny, $TSContext, $TSObject, AmplifySupportedService, exitOnNextTick, NotImplementedError } from 'amplify-cli-core';
import { UpdateApiRequest } from 'amplify-headless-interface';
import { printer } from 'amplify-prompts';
import inquirer from 'inquirer';
import * as path from 'path';
import { category } from '../../category-constants';
import { ApigwInputState } from './apigw-input-state';
import { getCfnApiArtifactHandler } from './cfn-api-artifact-handler';
import { addResource as addContainer, updateResource as updateContainer } from './containers-handler';
import { legacyAddResource } from './legacy-add-resource';
import {
  API_TYPE,
  getPermissionPolicies as getContainerPermissionPolicies,
  ServiceConfiguration,
} from './service-walkthroughs/containers-walkthrough';
import { datasourceMetadataFor, getServiceWalkthrough, serviceMetadataFor } from './utils/dynamic-imports';
import { editSchemaFlow } from './utils/edit-schema-flow';
import { serviceWalkthroughResultToAddApiRequest } from './utils/service-walkthrough-result-to-add-api-request';

export async function addAdminQueriesApi(
  context: $TSContext,
  apiProps: { apiName: string; functionName: string; authResourceName: string; dependsOn: $TSObject[] },
) {
  const apigwInputState = new ApigwInputState(context, apiProps.apiName);
  return apigwInputState.addAdminQueriesResource(apiProps);
}

export async function updateAdminQueriesApi(
  context: $TSContext,
  apiProps: { apiName: string; functionName: string; authResourceName: string; dependsOn: $TSObject[] },
) {
  const apigwInputState = new ApigwInputState(context, apiProps.apiName);
  // Check for migration

  if (!apigwInputState.cliInputsFileExists()) {
    await apigwInputState.migrateAdminQueries(apiProps);
  } else {
    return apigwInputState.updateAdminQueriesResource(apiProps);
  }
}

export async function console(context: $TSContext, service: string) {
  const { serviceWalkthroughFilename } = await serviceMetadataFor(service);
  const serviceWalkthroughSrc = path.join(__dirname, 'service-walkthroughs', serviceWalkthroughFilename);
  const { openConsole } = await import(serviceWalkthroughSrc);

  if (!openConsole) {
    const errMessage = 'Opening console functionality not available for this option';
    printer.error(errMessage);
    await context.usageData.emitError(new NotImplementedError(errMessage));
    exitOnNextTick(0);
  }

  return openConsole(context);
}

async function addContainerResource(context: $TSContext, service: string, options, apiType: API_TYPE) {
  const serviceWalkthroughFilename = 'containers-walkthrough.js';

  const serviceWalkthrough = await getServiceWalkthrough(serviceWalkthroughFilename);
  const serviceWalkthroughPromise: Promise<$TSAny> = serviceWalkthrough(context, apiType);

  return await addContainer(serviceWalkthroughPromise, context, category, service, options, apiType);
}

async function addNonContainerResource(context: $TSContext, service: string, options) {
  const serviceMetadata = await serviceMetadataFor(service);
  const { serviceWalkthroughFilename, defaultValuesFilename } = serviceMetadata;
  const serviceWalkthrough = await getServiceWalkthrough(serviceWalkthroughFilename);

  const serviceWalkthroughPromise: Promise<$TSAny> = serviceWalkthrough(context, serviceMetadata);
  switch (service) {
    case AmplifySupportedService.APPSYNC:
      const walkthroughResult = await serviceWalkthroughPromise;
      const askToEdit = walkthroughResult.askToEdit;
      const apiName = await getCfnApiArtifactHandler(context).createArtifacts(serviceWalkthroughResultToAddApiRequest(walkthroughResult));
      if (askToEdit) {
        await editSchemaFlow(context, apiName);
      }
      return apiName;
    case AmplifySupportedService.APIGW:
      const apigwInputState = new ApigwInputState(context);
      return apigwInputState.addApigwResource(serviceWalkthroughPromise, options);
    default:
      return legacyAddResource(serviceWalkthroughPromise, context, category, service, options);
  }
}

export async function addResource(context: $TSContext, service: string, options) {
  let useContainerResource = false;
  let apiType = API_TYPE.GRAPHQL;

  if (isContainersEnabled(context)) {
    switch (service) {
      case AmplifySupportedService.APPSYNC:
        useContainerResource = await isGraphQLContainer();
        apiType = API_TYPE.GRAPHQL;
        break;
      case AmplifySupportedService.APIGW:
        useContainerResource = await isRestContainer();
        apiType = API_TYPE.REST;
        break;
      default:
        throw new Error(`${service} not exists`);
    }
  }

  return useContainerResource
    ? addContainerResource(context, service, options, apiType)
    : addNonContainerResource(context, service, options);
}

function isContainersEnabled(context: $TSContext) {
  const { frontend } = context.amplify.getProjectConfig();
  if (frontend) {
    const { config: { ServerlessContainers = false } = {} } = context.amplify.getProjectConfig()[frontend] || {};

    return ServerlessContainers;
  }

  return false;
}

async function isGraphQLContainer(): Promise<boolean> {
  const { graphqlSelection } = await inquirer.prompt({
    name: 'graphqlSelection',
    message: 'Which service would you like to use',
    type: 'list',
    choices: [
      {
        name: AmplifySupportedService.APPSYNC,
        value: false,
      },
      {
        name: 'AWS Fargate (Container-based)',
        value: true,
      },
    ],
  });

  return graphqlSelection;
}

async function isRestContainer() {
  const { restSelection } = await inquirer.prompt({
    name: 'restSelection',
    message: 'Which service would you like to use',
    type: 'list',
    choices: [
      {
        name: 'API Gateway + Lambda',
        value: false,
      },
      {
        name: 'API Gateway + AWS Fargate (Container-based)',
        value: true,
      },
    ],
  });

  return restSelection;
}

export async function updateResource(context: $TSContext, category: string, service: string, options) {
  const allowContainers = options?.allowContainers ?? true;
  let useContainerResource = false;
  let apiType = API_TYPE.GRAPHQL;
  if (allowContainers && isContainersEnabled(context)) {
    const { hasAPIGatewayContainerResource, hasAPIGatewayLambdaResource, hasGraphQLAppSyncResource, hasGraphqlContainerResource } =
      await describeApiResourcesBySubCategory(context);

    switch (service) {
      case AmplifySupportedService.APPSYNC:
        if (hasGraphQLAppSyncResource && hasGraphqlContainerResource) {
          useContainerResource = await isGraphQLContainer();
        } else if (hasGraphqlContainerResource) {
          useContainerResource = true;
        } else {
          useContainerResource = false;
        }
        apiType = API_TYPE.GRAPHQL;
        break;
      case AmplifySupportedService.APIGW:
        if (hasAPIGatewayContainerResource && hasAPIGatewayLambdaResource) {
          useContainerResource = await isRestContainer();
        } else if (hasAPIGatewayContainerResource) {
          useContainerResource = true;
        } else {
          useContainerResource = false;
        }
        apiType = API_TYPE.REST;
        break;
      default:
        throw new Error(`${service} not exists`);
    }
  }

  return useContainerResource ? updateContainerResource(context, category, service, apiType) : updateNonContainerResource(context, service);
}

async function describeApiResourcesBySubCategory(context: $TSContext) {
  const { allResources } = await context.amplify.getResourceStatus();
  const resources = allResources.filter(resource => resource.category === category && resource.mobileHubMigrated !== true);

  let hasAPIGatewayContainerResource = false;
  let hasAPIGatewayLambdaResource = false;
  let hasGraphQLAppSyncResource = false;
  let hasGraphqlContainerResource = false;

  resources.forEach(resource => {
    hasAPIGatewayContainerResource =
      hasAPIGatewayContainerResource || (resource.service === 'ElasticContainer' && resource.apiType === API_TYPE.REST);

    hasAPIGatewayLambdaResource = hasAPIGatewayLambdaResource || resource.service === AmplifySupportedService.APIGW;

    hasGraphQLAppSyncResource = hasGraphQLAppSyncResource || resource.service === AmplifySupportedService.APPSYNC;

    hasGraphqlContainerResource =
      hasGraphqlContainerResource || (resource.service === 'ElasticContainer' && resource.apiType === API_TYPE.GRAPHQL);
  });

  return {
    hasAPIGatewayLambdaResource,
    hasAPIGatewayContainerResource,
    hasGraphQLAppSyncResource,
    hasGraphqlContainerResource,
  };
}

async function updateContainerResource(context: $TSContext, category: string, service: string, apiType: API_TYPE) {
  const serviceWalkthroughFilename = 'containers-walkthrough';
  const serviceWalkthroughSrc = path.join(__dirname, 'service-walkthroughs', serviceWalkthroughFilename);
  const { updateWalkthrough } = await import(serviceWalkthroughSrc);

  if (!updateWalkthrough) {
    const errMessage = 'Update functionality not available for this option';
    printer.error(errMessage);
    await context.usageData.emitError(new NotImplementedError(errMessage));
    exitOnNextTick(0);
  }

  const updateWalkthroughPromise: Promise<ServiceConfiguration> = updateWalkthrough(context, apiType);

  updateContainer(updateWalkthroughPromise, context, category);
}

async function updateNonContainerResource(context: $TSContext, service: string) {
  const serviceMetadata = await serviceMetadataFor(service);
  const { defaultValuesFilename, serviceWalkthroughFilename } = serviceMetadata;
  const serviceWalkthroughSrc = path.join(__dirname, 'service-walkthroughs', serviceWalkthroughFilename);
  const { updateWalkthrough } = await import(serviceWalkthroughSrc);

  if (!updateWalkthrough) {
    const errMessage = 'Update functionality not available for this option';
    printer.error(errMessage);
    await context.usageData.emitError(new NotImplementedError(errMessage));
    exitOnNextTick(0);
  }

  const updateWalkthroughPromise: Promise<UpdateApiRequest> = updateWalkthrough(context, defaultValuesFilename, serviceMetadata);

  switch (service) {
    case AmplifySupportedService.APPSYNC:
      return updateWalkthroughPromise.then(getCfnApiArtifactHandler(context).updateArtifacts);
    default:
      const apigwInputState = new ApigwInputState(context);
      return apigwInputState.updateApigwResource(updateWalkthroughPromise);
  }
}

export async function migrateResource(context: $TSContext, projectPath: string, service: string, resourceName: string) {
  if (service === 'ElasticContainer') {
    return migrateResourceContainer(context, projectPath, service, resourceName);
  } else {
    return migrateResourceNonContainer(context, projectPath, service, resourceName);
  }
}

async function migrateResourceContainer(context: $TSContext, projectPath: string, service: string, resourceName: string) {
  printer.info(`No migration required for ${resourceName}`);
  return;
}

async function migrateResourceNonContainer(context: $TSContext, projectPath: string, service: string, resourceName: string) {
  const serviceMetadata = await serviceMetadataFor(service);
  const { serviceWalkthroughFilename } = serviceMetadata;
  const serviceWalkthroughSrc = path.join(__dirname, 'service-walkthroughs', serviceWalkthroughFilename);
  const { migrate } = await import(serviceWalkthroughSrc);

  if (!migrate) {
    printer.info(`No migration required for ${resourceName}`);
    return;
  }

  return await migrate(context, projectPath, resourceName);
}

export async function addDatasource(context: $TSContext, category, datasource) {
  const serviceMetadata = await datasourceMetadataFor(datasource);
  const { serviceWalkthroughFilename } = serviceMetadata;
  return (await getServiceWalkthrough(serviceWalkthroughFilename))(context, serviceMetadata);
}

export async function getPermissionPolicies(context: $TSContext, service: string, resourceName: string, crudOptions) {
  if (service === 'ElasticContainer') {
    return getPermissionPoliciesContainer(context, service, resourceName, crudOptions);
  } else {
    return getPermissionPoliciesNonContainer(service, resourceName, crudOptions);
  }
}

async function getPermissionPoliciesContainer(context: $TSContext, service: string, resourceName: string, crudOptions) {
  return getContainerPermissionPolicies(context, service, resourceName, crudOptions);
}

async function getPermissionPoliciesNonContainer(service: string, resourceName: string, crudOptions: string[]) {
  const serviceMetadata = await serviceMetadataFor(service);
  const { serviceWalkthroughFilename } = serviceMetadata;
  const serviceWalkthroughSrc = path.join(__dirname, 'service-walkthroughs', serviceWalkthroughFilename);
  const { getIAMPolicies } = await import(serviceWalkthroughSrc);

  if (!getIAMPolicies) {
    printer.info(`No policies found for ${resourceName}`);
    return;
  }

  return getIAMPolicies(resourceName, crudOptions);
}
