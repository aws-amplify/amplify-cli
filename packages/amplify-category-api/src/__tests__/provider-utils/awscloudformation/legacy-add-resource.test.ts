import { legacyAddResource } from '../../../provider-utils/awscloudformation/legacy-add-resource';
import { category } from '../../../category-constants';

jest.mock('fs-extra');

describe('legacy add resource', () => {
  const contextStub = {
    amplify: {
      pathManager: {
        getBackendDirPath: jest.fn(_ => 'mock/backend/path'),
      },
      updateamplifyMetaAfterResourceAdd: jest.fn(),
      copyBatch: jest.fn(),
    },
  };

  it('sets policy resource name in paths object before copying template', async () => {
    const stubWalkthroughPromise: Promise<any> = Promise.resolve({
      answers: {
        resourceName: 'mockResourceName',
        paths: [
          {
            name: '/some/{path}/with/{params}',
          },
          {
            name: 'another/path/without/params',
          },
        ],
      },
    });
    await legacyAddResource(stubWalkthroughPromise, contextStub, category, 'API Gateway', {});
    expect(contextStub.amplify.copyBatch.mock.calls[0][2]).toMatchSnapshot();
  });
});
