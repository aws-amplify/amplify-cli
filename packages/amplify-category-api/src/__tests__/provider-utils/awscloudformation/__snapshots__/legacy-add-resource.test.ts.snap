// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`legacy add resource sets policy resource name in paths object before copying template 1`] = `
Object {
  "paths": Array [
    Object {
      "name": "/some/{path}/with/{params}",
      "policyResourceName": "/some/*/with/*",
    },
    Object {
      "name": "another/path/without/params",
      "policyResourceName": "another/path/without/params",
    },
  ],
  "resourceName": "mockResourceName",
}
`;
